spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTransactionManager.java

```diff
@@ -0,0 +1,240 @@
+							+ "TransactionDefinition.PROPAGATION_REQUIRED");
+		}
+		Tx tx = (Tx) transactionObject;
+		Transaction transaction = this.datastore.newTransaction();
+		if (transactionDefinition.isReadOnly()) {
+			tx.transaction = new Transaction() {
+				@Override
+				public Entity get(Key key) {
+					return transaction.get(key);
+				}
+
+				@Override
+				public Iterator<Entity> get(Key... key) {
+					return transaction.get(key);
+				}
+
+				@Override
+				public List<Entity> fetch(Key... keys) {
+					return transaction.fetch(keys);
+				}
+
+				@Override
+				public <T> QueryResults<T> run(Query<T> query) {
+					return transaction.run(query);
+				}
+
+				@Override
+				public void addWithDeferredIdAllocation(FullEntity<?>... entities) {
+					throw new UnsupportedOperationException(
+							"The Cloud Datastore transaction is in read-only mode.");
+				}
+
+				@Override
+				public Entity add(FullEntity<?> entity) {
+					throw new UnsupportedOperationException(
+							"The Cloud Datastore transaction is in read-only mode.");
+				}
+
+				@Override
+				public List<Entity> add(FullEntity<?>... entities) {
+					throw new UnsupportedOperationException(
+							"The Cloud Datastore transaction is in read-only mode.");
+				}
+
+				@Override
+				public void update(Entity... entities) {
+					throw new UnsupportedOperationException(
+							"The Cloud Datastore transaction is in read-only mode.");
+				}
+
+				@Override
+				public void delete(Key... keys) {
+					throw new UnsupportedOperationException(
+							"The Cloud Datastore transaction is in read-only mode.");
+				}
+
+				@Override
+				public void putWithDeferredIdAllocation(FullEntity<?>... entities) {
+					throw new UnsupportedOperationException(
+							"The Cloud Datastore transaction is in read-only mode.");
+				}
+
+				@Override
+				public Entity put(FullEntity<?> entity) {
+					throw new UnsupportedOperationException(
+							"The Cloud Datastore transaction is in read-only mode.");
+				}
+
+				@Override
+				public List<Entity> put(FullEntity<?>... entities) {
+					throw new UnsupportedOperationException(
+							"The Cloud Datastore transaction is in read-only mode.");
+				}
+
+				@Override
+				public Response commit() {
+					return transaction.commit();
+				}
+
+				@Override
+				public void rollback() {
+					transaction.rollback();
+				}
+
+				@Override
+				public boolean isActive() {
+					return transaction.isActive();
+				}
+
+				@Override
+				public Datastore getDatastore() {
+					return transaction.getDatastore();
+				}
+
+				@Override
+				public ByteString getTransactionId() {
+					return transaction.getTransactionId();
+				}
+			};
+		}
+		else {
+			tx.transaction = transaction;
+		}
+	}
+
+	@Override
+	protected void doCommit(DefaultTransactionStatus defaultTransactionStatus)
+			throws TransactionException {
+		Tx tx = (Tx) defaultTransactionStatus.getTransaction();
+		try {
+			if (tx.transaction.isActive()) {
```

Comment:
Add debug line about when it's not active. check doc for the parent if we should check active or not